cryptol-remote-api/python/cryptol/synchronous.py

```diff
@@ -0,0 +1,283 @@
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`, and return up to
+        `count` solutions.
+        """
+        if isinstance(solver, OfflineSolver):
+            res = self.connection.sat_raw(expr, solver, count).result()
+            if res['result'] == 'offline':
+                return OfflineSmtQuery(res['query'])
+            else:
+                raise ValueError("Expected an offline SMT result, got: " + str(res))
+        elif isinstance(solver, OnlineSolver):
+            res = self.connection.sat_raw(expr, solver, count).result()
+            if res['result'] == 'unsatisfiable':
+                return Unsatisfiable()
+            elif res['result'] == 'satisfied':
+                return Satisfiable([[from_cryptol_arg(arg['expr']) for arg in m] for m in res['models']])
+            else:
+                raise ValueError("Unexpected 'sat' SMT result: " + str(res))
+        else:
+            raise ValueError("Unknown solver type: " + str(solver))
+
+    @overload
+    def prove(self, expr : Any, solver : OnlineSolver = Z3) -> Union[Qed, Counterexample]: ...
+    @overload
+    def prove(self, expr : Any, solver : OfflineSolver) -> OfflineSmtQuery: ...
+
+    def prove(self, expr : Any, solver : Solver = Z3) -> Union[Qed, Counterexample, OfflineSmtQuery]:
+        """Check the validity of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`.
+        """
+        if isinstance(solver, OfflineSolver):
+            res = self.connection.prove_raw(expr, solver).result()
+            if res['result'] == 'offline':
+                return OfflineSmtQuery(res['query'])
+            else:
+                raise ValueError("Expected an offline SMT result, got: " + str(res))
+        elif isinstance(solver, OnlineSolver):
+            res = self.connection.prove_raw(expr, solver).result()
+            if res['result'] == 'unsatisfiable':
+                return Qed()
+            elif res['result'] == 'invalid':
+                return Counterexample(res['counterexample type'], [from_cryptol_arg(arg['expr']) for arg in res['counterexample']])
+            else:
+                raise ValueError("Unexpected 'prove' SMT result: " + str(res))
+        else:
+            raise ValueError("Unknown solver type: " + str(solver))
+
+    @overload
+    def safe(self, expr : Any, solver : OnlineSolver = Z3) -> Union[Safe, Counterexample]: ...
+    @overload
+    def safe(self, expr : Any, solver : OfflineSolver) -> OfflineSmtQuery: ...
+
+    def safe(self, expr : Any, solver : Solver = Z3) -> Union[Safe, Counterexample, OfflineSmtQuery]:
+        """Check via an external SMT solver that the given term is safe for all inputs,
+        which means it cannot encounter a run-time error.
```

Comment:
We should add a note in this docstring regarding being able to boolean test `Safe ` and `Counterexample ` (i.e., that the latter is falsy).